camel/models/open_source_model.py

```diff
@@ -0,0 +1,102 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import os
+from types import GeneratorType
+from typing import Any, Dict, List, Optional
+
+from camel.messages import OpenAIMessage
+from camel.models import BaseModelBackend
+from camel.typing import ModelType
+
+
+class OpenSourceModel(BaseModelBackend):
```

Comment:
I agree with that this class has quite a lot in common with `OpenAIModel` but it introduces more checks currently. And if we need to do further modifications on the class for open-source models, it will be more convenient to have a separate class for them. I can currently modify it to inherit from the `OpenAIModel` to avoid some unnecessary copy-pastes while not making the two classes too complex.

Reply:
> And if we need to do further modifications on the class for open-source models, it will be more convenient to have a separate class for them.

The `BaseModelBackend` level is mainly related to APIs, so putting too much logic here may not be a good idea.

> I can currently modify it to inherit from the OpenAIModel to avoid some unnecessary copy-pastes while not making the two classes too complex.

This solution is better. In advance, how about creating a class for all OpenAI API models, including all open source models. Then let `OpenAIModel` inherit it, which is conceptually more clear than letting `OpenSourceModel` inherit.

Reply:
The main diffs are checking model_type, model_path, base_url. All of these checks can be removed if we correctly put these information in `model_config`.

Reply:
Maybe we can find a better way to remove code duplication.